File: packages/needle/src/common/trie.ts
export interface TrieNode {
  parent: TrieNode | undefined;
  children: Map<number, TrieNode>; // Unicode code point -> child node
  tokenIds: number[];
  subTreeTokenIds: number[]; // Empty on root. Will Uint16Array be faster?
}

export const traverseTrieStep = (node: TrieNode | undefined, codePoint: string, ignorableCodePoints?: RegExp) =>
  node?.children.get(codePoint.codePointAt(0)!) ?? (ignorableCodePoints?.test(codePoint) ? node : undefined);
export const traverseTrie = (node: TrieNode | undefined, text: string, ignorableCodePoints?: RegExp) => {
  if (!node) return;
  for (const codePoint of text) {
    node = traverseTrieStep(node, codePoint, ignorableCodePoints);
    if (!node) return;
  }
  return node;
};

File: dotnet/MaigoLabs.NeedLe.Common/Trie.cs
namespace MaigoLabs.NeedLe.Common;

public class TrieNode
{
    public required TrieNode? Parent { get; set; }
    public required Dictionary<int, TrieNode> Children { get; set; } // Unicode code point -> child node
    public required List<int> TokenIds { get; set; }
    public required List<int> SubTreeTokenIds { get; set; } // Empty on root.
}

public static class TrieNodeExtensions
{
    public static TrieNode? TraverseStep(this TrieNode? node, int codePoint, bool isIgnorable = false) =>
        (node?.Children.TryGetValue(codePoint, out var child) ?? false)
            ? child
            : isIgnorable ? node : null;

    public static TrieNode? Traverse(this TrieNode? node, int[] codePoints, bool isIgnorable = false)
    {
        if (node == null) return null;
        foreach (var codePoint in codePoints)
        {
            node = node?.TraverseStep(codePoint, isIgnorable);
            if (node == null) return null;
        }
        return node;
    }

    public static List<int> GetTokenIds(this TrieNode? node, bool includeSubTree = false) =>
        (includeSubTree ? node?.SubTreeTokenIds : node?.TokenIds) ?? [];

    public static bool IsTokenExactMatch(this TrieNode? node, int tokenId) => node?.TokenIds.Contains(tokenId) ?? false;
}
